cpp/src/arrow/flight/flight_test.cc

```diff
@@ -1077,5 +1327,169 @@ TEST_F(TestMetadata, DoPutReadMetadata) {
+}
+
+// For each method, make sure that the client middleware received
+// headers from the server and that the proper method enum value was
+// passed to the interceptor
+TEST_F(TestPropagatingMiddleware, ListFlights) {
+  client_middleware_->Reset();
+  std::unique_ptr<FlightListing> listing;
+  const Status status = client_->ListFlights(&listing);
+  ASSERT_RAISES(NotImplemented, status);
+  ValidateStatus(status, FlightMethod::ListFlights);
+}
+
+TEST_F(TestPropagatingMiddleware, GetFlightInfo) {
+  client_middleware_->Reset();
+  auto descr = FlightDescriptor::Path({"examples", "ints"});
+  std::unique_ptr<FlightInfo> info;
+  const Status status = client_->GetFlightInfo(descr, &info);
+  ASSERT_RAISES(NotImplemented, status);
+  ValidateStatus(status, FlightMethod::GetFlightInfo);
+}
+
+TEST_F(TestPropagatingMiddleware, GetSchema) {
+  client_middleware_->Reset();
+  auto descr = FlightDescriptor::Path({"examples", "ints"});
+  std::unique_ptr<SchemaResult> result;
+  const Status status = client_->GetSchema(descr, &result);
+  ASSERT_RAISES(NotImplemented, status);
+  ValidateStatus(status, FlightMethod::GetSchema);
+}
+
+TEST_F(TestPropagatingMiddleware, ListActions) {
+  client_middleware_->Reset();
+  std::vector<ActionType> actions;
+  const Status status = client_->ListActions(&actions);
+  ASSERT_RAISES(NotImplemented, status);
+  ValidateStatus(status, FlightMethod::ListActions);
+}
+
+TEST_F(TestPropagatingMiddleware, DoGet) {
+  client_middleware_->Reset();
+  Ticket ticket1{"ARROW-5095-fail"};
+  std::unique_ptr<FlightStreamReader> stream;
+  Status status = client_->DoGet(ticket1, &stream);
+  ASSERT_RAISES(NotImplemented, status);
+  ValidateStatus(status, FlightMethod::DoGet);
+}
+
+TEST_F(TestPropagatingMiddleware, DoPut) {
+  client_middleware_->Reset();
+  auto descr = FlightDescriptor::Path({"ints"});
+  auto a1 = ArrayFromJSON(int32(), "[4, 5, 6, null]");
+  auto schema = arrow::schema({field("f1", a1->type())});
+
+  std::unique_ptr<FlightStreamWriter> stream;
+  std::unique_ptr<FlightMetadataReader> reader;
+  ASSERT_OK(client_->DoPut(descr, schema, &stream, &reader));
+  const Status status = stream->Close();
+  ASSERT_RAISES(NotImplemented, status);
+  ValidateStatus(status, FlightMethod::DoPut);
+}
+
+// Validate the implementation of the call headers wrapper
+TEST_F(TestPropagatingMiddleware, CallHeaders) {
+  std::string header1 = "x-header";
+  std::string header1_value1 = "value";
+  std::string header1_value2 = "value";
+  std::string header2 = "x-header-two";
+  std::string header2_value1 = "value-two";
+
+  internal::GrpcMetadataMap headers;
+  headers.insert({header1, header1_value1});
+  headers.insert({header2, header2_value1});
+  headers.insert({header1, header1_value2});
+
+  internal::GrpcCallHeaders incoming_headers(&headers);
+
+  ASSERT_EQ(incoming_headers.cbegin(), incoming_headers.cbegin());
+  ASSERT_EQ(incoming_headers.cend(), incoming_headers.cend());
```

Comment:
I'll clarify inline, but I ran into an issue where this wasn't true because I had mis-implemented the underlying iterator (and didn't realize it until I was testing the Python bindings). (Side note, `HeaderIterator` is kind of kludgy, if you know of a better way in C++ to accomplish what I'm trying there...)